### PR TITLE
RBAC support for flow engine and auth code flow

### PR DESCRIPTION
**Pull request by senthalan**
Part of https://github.com/asgardeo/thunder/issues/510

This pull request introduces a new pluggable authorization service to the backend, centered around an RBAC (Role-Based Access Control) engine that delegates permission checks to the role service. The changes include the implementation of the core authorization service, the RBAC engine, supporting models and error handling, as well as integration into the service initialization process and the authentication flow. Comprehensive unit tests for the RBAC engine are also included.

**Authorization Service Implementation**

- Added a new `AuthorizationServiceInterface` and its implementation in `service.go`, providing a method to determine which permissions a user or group is authorized for. The service delegates authorization logic to a pluggable engine and handles logging and error responses.
- Defined request and response models for authorization checks in `model.go`.
- Introduced error constants for authorization failures in `errorconstants.go`.
- Created an internal `authorizationEngine` interface in `engine.go` to allow for different pluggable authorization engines (e.g., RBAC, ABAC).

**RBAC Engine Implementation and Testing**

- Implemented the RBAC engine in `engines/rbacengine.go`, which delegates permission checks to the role service.
- Added comprehensive unit tests for the RBAC engine in `engines/rbacengine_test.go`, covering success, error, and edge cases.

**Service Initialization and Integration**

- Added an `Initialize` function in `init.go` to set up the authorization service with the RBAC engine and expose a singleton getter.
- Integrated the new authorization service into the backend service initialization in `servicemanager.go`, ensuring it is initialized with the role service. [[1]](diffhunk://#diff-36914807f52c25f79a7d09a3a5cca10a8adb6df10e3df582e89ea4c713143216R26) [[2]](diffhunk://#diff-36914807f52c25f79a7d09a3a5cca10a8adb6df10e3df582e89ea4c713143216L56-R58)
- Updated mockery configuration to generate mocks for the new `authz` and `role` packages, supporting unit testing.

**Authentication Flow Integration**

- Updated the authentication flow configuration (`auth_flow_config_basic.json`) to include an `authorization_check` step after basic authentication, using the new `AuthorizationExecutor`.

### CODE REVIEW

File: backend/internal/role/store.go [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+) ([]string, error) {
+	dbClient, err := s.getIdentityDBClient()
+	if err != nil {
+		return nil, err
+	}
+
+	// Handle nil groupIDs slice
+	if groupIDs == nil {
+		groupIDs = []string{}
+	}
```
[the remark this answers is not in the document]
Comment (senthalan) [reply backend/internal/role/store.go:449]: This changes are removed from this PR

File: backend/internal/role/storeconstants.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+)
+
+// buildAuthorizedPermissionsQuery constructs a database-specific query to retrieve authorized permissions
+// for a user and/or groups from their assigned roles.
+// It builds separate queries for PostgreSQL and SQLite to handle array parameters correctly.
+func buildAuthorizedPermissionsQuery(
+	userID string,
+	groupIDs []string,
+	requestedPermissions []string,
+) (dbmodel.DBQuery, []interface{}) {
```
[the remark this answers is not in the document]
Comment (senthalan) [reply backend/internal/role/storeconstants.go:136]: This changes are removed from this PR

File: backend/internal/executor/authz/authzexecutor.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+	// Call authorization service
+	authzReq := authzsvc.GetAuthorizedPermissionsRequest{
+		UserID:               userID,
+		GroupIDs:             groupIDs,
+		RequestedPermissions: requestedPerms,
+	}
+
+	authzResp, svcErr := a.authzService.GetAuthorizedPermissions(authzReq)
+	if svcErr != nil {
+		logger.Error("Authorization service call failed", log.String("error", svcErr.Error))
```
[the remark this answers is not in the document]
Comment (senthalan) [reply backend/internal/executor/authz/authzexecutor.go:107]: Updated to fail the authentication flow when there is an error returned from the authorization service

File: backend/internal/executor/authassert/authassertexecutor.go [hunk 4]
```diff
@@ -176,6 +176,11 @@ func (a *AuthAssertExecutor) generateAuthAssertion(ctx *flowmodel.NodeContext, l
 		jwtClaims["assurance"] = assertionResult.Context
 	}
 
+	// Include authorized permissions in JWT if present in runtime data
```
[the remark this answers is not in the document]
Comment (senthalan) [reply backend/internal/executor/authassert/authassertexecutor.go:179]: Improved the comment

File: tests/integration/oauth/authz/authz_scope_test.go [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+}
+
+func TestOAuthAuthzScopeTestSuite(t *testing.T) {
+	suite.Run(t, new(OAuthAuthzScopeTestSuite))
+}
+
+func (ts *OAuthAuthzScopeTestSuite) SetupSuite() {
+	var err error
+
+	// Setup HTTP client
```
[the remark this answers is not in the document]
Comment (senthalan) [reply tests/integration/oauth/authz/authz_scope_test.go:81]: This is acceptable for integration tests

File: backend/internal/oauth/oauth2/authz/handler.go [hunk 6]
```diff
@@ -258,8 +264,9 @@ func (ah *authorizeHandler) handleAuthorizationResponseFromEngine(msg *OAuthMess
 		return
 	}
 
-	// TODO: Do user authorization.
-	//  Should validate for the scopes as well.
+	authorizedScopes := attributes["authorized_permissions"]
+	// Overwrite the non oidc scopes in session data with the authorized scopes from the assertion.
+	sessionData.OAuthParameters.NonOIDCScopes = utils.ParseStringArray(authorizedScopes, " ")
```
[the remark this answers is not in the document]
Comment (senthalan) [reply backend/internal/oauth/oauth2/authz/handler.go]: This is intentional, We need to override the NonOIDCScopes(PermissionScopes) with the authorizedScopes always.

File: backend/internal/executor/authz/authzexecutor.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+		log.String(log.LoggerKeyFlowID, ctx.FlowID))
+	logger.Debug("Executing authorization executor")
+
+	execResp := &flowmodel.ExecutorResponse{
+		RuntimeData: make(map[string]string),
+	}
+
+	if !ctx.AuthenticatedUser.IsAuthenticated {
+		execResp.Status = flowconst.ExecFailure
+		execResp.FailureReason = "User is not authenticated"
```
[the remark this answers is not in the document]
Comment (senthalan) [reply backend/internal/executor/authz/authzexecutor.go:74]: This is expected. We can't perform authorization when the user is not authenticated

File: backend/internal/system/error/serviceerror/error.go [hunk 8]
```diff
@@ -50,3 +50,10 @@ func CustomServiceError(svcError ServiceError, errorDesc string) *ServiceError {
 	}
 	return err
 }
+
+// Server errors
+var (
+	// EncodingError is the error returned when encoding the response.
+	ErrorEncodingError = "{Code: \"ENC-5000\",Error: \"Encoding error\"," +
+		"ErrorDescription: \"An error occurred while encoding the response\"}"
```
[the remark this answers is not in the document]
Comment (senthalan) [reply backend/internal/system/error/serviceerror/error.go:58]: this change is not part of this PR anymore

File: tests/integration/testutils/apiutils.go [hunk 9]
```diff
@@ -494,3 +494,201 @@ func FindUserByAttribute(key, value string) (*User, error) {
...
+	if err := json.Unmarshal(respBody, &createdRole); err != nil {
+		return "", fmt.Errorf("failed to unmarshal role response: %w", err)
+	}
+
+	return createdRole.ID, nil
+}
+
+// DeleteRole deletes a role by ID
+// It first removes all assignments before deleting the role, as roles with assignments cannot be deleted
+func DeleteRole(roleID string) error {
```
[the remark this answers is not in the document]
Comment (senthalan) [reply tests/integration/testutils/apiutils.go:593]: Removed the comment

File: backend/cmd/server/repository/resources/graphs/auth_flow_config_basic.json [hunk 10]
```diff
@@ -8,6 +8,16 @@
             "executor": {
                 "name": "BasicAuthExecutor"
             },
+            "next": [
```
Comment (rajithacharith) [inline backend/cmd/server/repository/resources/graphs/auth_flow_config_basic.json:11]: This means authorization will not be available by default. This has to be added for the flow to use RBAC support?

File: backend/internal/authz/engines/rbacengine.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// Package engines provides authorization engine implementations.
+// It includes various authorization engines such as RBAC (Role-Based Access Control)
+// that delegate authorization decisions to the appropriate services.
+package engines
```
Comment (ThaminduDilshan) [inline backend/internal/authz/engines/rbacengine.go]:
```suggestion
package engine
```

Shall we use engine?

File: backend/cmd/server/repository/resources/graphs/auth_flow_config_basic.json (hunk 10 above)
Comment (senthalan) [reply backend/cmd/server/repository/resources/graphs/auth_flow_config_basic.json:11]: In the current state, this is the behaviour. I modelled it like the application developer can turn on or off authorization check.

But can revisit it and engage the authorization by default.

File: backend/internal/authz/service.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+		request.UserID,
+		request.GroupIDs,
+		request.RequestedPermissions,
+	)
+	if err != nil {
+		logger.Error("Authorization evaluation failed",
+			log.String("userID", request.UserID),
+			log.Int("groupCount", len(request.GroupIDs)),
+			log.Error(err))
+		return nil, &ErrorAuthorizationFailed
```
Comment (ThaminduDilshan) [inline backend/internal/authz/service.go:85]: Does this have to be a server error always?

File: backend/internal/authz/engine.go [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package authz
+
+// authorizationEngine is the internal interface for authorization engines.
+// This interface is NOT exported and is used internally by the authorization service.
+// Different engines can be plugged in (RBAC, ABAC, ReBAC, Custom) by implementing this interface.
+type authorizationEngine interface {
```
Comment (ThaminduDilshan) [inline backend/internal/authz/engine.go]: I'd suggest to make this public and let rbac engine implement this. Currently we cannot guarantee engine impl implements the authorizationEngine interface right?

File: backend/internal/authz/service_test.go [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
+)
+
+// mockAuthorizationEngine is a mock implementation of authorizationEngine for testing.
+type mockAuthorizationEngine struct {
```
Comment (ThaminduDilshan) [inline backend/internal/authz/service_test.go]: Can't we use mockery generated mocks rather than writing custom mock implementations?

File: backend/internal/executor/authz/authzexecutor.go [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+			id, name, flowconst.ExecutorTypeUtility,
+			[]flowmodel.InputData{}, []flowmodel.InputData{}, properties,
+		),
+		authzService: authzsvc.GetAuthorizationService(),
+	}
+}
+
+// Execute executes the authorization logic by determining required permissions based on context,
+// calling the authorization service, and storing authorized permissions in runtime data.
+// This executor never fails the flow - it always returns ExecComplete status.
```
Comment (thiva-k) [inline backend/internal/executor/authz/authzexecutor.go]: This can fail the flow if Authz service call fails/ returns an error right?

File: backend/internal/authz/service.go (hunk 12 above)
Comment (senthalan) [reply backend/internal/authz/service.go:85]: yeah, There is no possibility of a client side error here. If the user provided invalid permissions or something like that, the authorized_permission list will be returned as empty. 
 The errors will be due to the server side.

**Review by ThaminduDilshan — APPROVED** [review]
